### PR TITLE
Fix: Avoid compromising accuracy by prematurely applying half-precision.

### PR DESCRIPTION
**Pull request by LiuXTao**
I noticed that when running the **llama** model, **the rope embedding is first converted to half-precision and then processed using the cosine function. This approach differs significantly from the method of applying the cosine function first and then converting to half-precision**.

In contrast, both the "**transformers**" and "**xformers**" libraries follow the practice of applying the cosine function first and then converting to half-precision. You can find examples of this in the following code repositories:

Transformers: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py
Xformers: https://github.com/facebookresearch/xformers/blob/main/xformers/components/positional_embedding/rotary.py

For instance, let's consider an example where the seq-length is set to 2048, and a data sample is randomly selected from the result of the rope_embedding. The data sample, denoted as A, looks like this:

A = [5.1000000000000000e+02, 4.4164181518554688e+02, 3.8244604492187500e+02, ..., 7.8536249697208405e-02, 6.8009600043296814e-02, 5.8893878012895584e-02]

**Method 1: A.half().cos()**:
If we apply the half-precision conversion to A first and then take the cosine, the resulting values will be:
[0.4870605468750000, -0.3486328125000000, 0.7148437500000000, ..., 0.9970703125000000, 0.9975585937500000, 0.9980468750000000]

**Method 2: A.cos().half()**:
On the other hand, if we apply the cosine function to A first and then convert to half-precision, the resulting values will be:
[0.4870605468750000, -0.2454833984375000, 0.6762695312500000, ..., 0.9970703125000000, 0.9975585937500000, 0.9980468750000000]

**It is evident that the results obtained by the two methods are different**. In the given example, (4.4164181518554688e+02).half().cos() equals -0.3486328125000000, while (4.4164181518554688e+02).cos().half() equals -0.2454833984375000. **The difference between the two values is 0.1.**

Therefore, normally **it is advisable to take the cosine first and then convert to half-precision, which can help avoid such significant loss in accuracy**.

### CODE REVIEW

**Comment by LiuXTao** [comment]
@microsoft-github-policy-service agree

**Comment by conglongli** [comment]
@LydiaXiaohongLi (https://github.com/microsoft/Megatron-DeepSpeed/pull/166) and @nrailgun (https://github.com/microsoft/Megatron-DeepSpeed/pull/177), could you give some feedback about this PR? Thanks.

**Comment by nrailgun** [comment]
> @LydiaXiaohongLi (#166) and @nrailgun (#177), could you give some feedback about this PR? Thanks.

I ran the following code

```python
import torch

dim = 4096
max_seq_len = 2048

inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2).float() / dim))
inv_freq = inv_freq.cuda()
assert inv_freq.dtype == torch.float32

seq = torch.arange(max_seq_len)
seq = seq.cuda()

freqs = torch.einsum('i , j -> i j', seq.type_as(inv_freq), inv_freq)
assert freqs.dtype == torch.float32

resultA = freqs.cos().half()
assert resultA.dtype == torch.half

resultB = freqs.half().cos()
assert resultB.dtype == torch.half

print(resultA)
print(resultB)
```

and it turned out that `resultA` and `resultB` are very different.

image

I suggest we should use `freq.cos().half()` instead, as `transformers` lib did.

**Comment by LydiaXiaohongLi** [comment]
Thanks for pointing out, agree with the change.

**Review by conglongli — APPROVED** [review]
